clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7443,6 +7443,44 @@ static bool checkForDuplicateAttribute(Sema &S, Decl *D,
   return false;
 }
 
+// Checks if FPGA memory attributes apply on valid variables.
+// Returns true if an error occured.
+static bool SYCLIntelConstLocalStaticAgentMemVar(Sema &S, Decl *D) {
+  if (const auto *VD = dyn_cast<VarDecl>(D)){
+    if (!(isa<FieldDecl>(D) ||
+          (VD->getKind() != Decl::ImplicitParam &&
+           VD->getKind() != Decl::NonTypeTemplateParm &&
+	   (VD->getStorageClass() == SC_Static ||
+            VD->hasLocalStorage() ||
+            (VD->getKind() != Decl::ParmVar &&
+             (S.isTypeDecoratedWithDeclAttribute<SYCLDeviceGlobalAttr>(VD->getType()) ||
+              VD->getType().isConstQualified() ||
+              VD->getType().getAddressSpace() ==
+                  LangAS::opencl_constant)))))) {
```

Comment:
With the way that the original ``SYCLIntelLocalStaticAgentMemVar`` is combined with ``SYCLIntelConstVar``, I don't think there is any difference between this routine and the routine below.   So maybe we can just combine it.  If so keep ``SYCLIntelConstOrLocalOrStaticVar`` - I like the formatting in that one better.

(I am sorry Soumi; this may be why you had them combined in an earlier iteration.   I just didn't understand it and thought they were two different checks.)

Reply:
No Problem @premanandrao. I have combined it. There is one difference i noticed while doing more investigation and testing.
```
VD->getKind() != Decl::ParmVar &&
```

I have added this existing restriction for impacted attributes:
```
// expected-error@+1{{'singlepump' attribute only applies to constant variables, local variables, static variables, non-static data members and device_global variables}}
void attr_on_func_arg7([[intel::singlepump]] int pc7) {}
// expected-error@+1{{'doublepump' attribute only applies to constant variables, local variables, static variables, non-static data members and device_global variables}}
void attr_on_func_arg8([[intel::doublepump]] int pc8) {}
// expected-error@+1{{'fpga_register' attribute only applies to constant variables, local variables, static variables, non-static data members and device_global variables}}
void attr_on_func_arg9([[intel::fpga_register]] int pc9) {}
// expected-error@+1{{'merge' attribute only applies to constant variables, local variables, static variables, non-static data members and device_global variables}}
void attr_on_func_arg10([[intel::merge("mrg1", "width")]] int pc10) {}

```